File: scripts/proof-workers/orchestrator/issues.py
"""Issue fetching, caching, and dependency resolution."""

from __future__ import annotations

import subprocess
from typing import Optional

from .config import RunConfig
from .models import Issue
from .state import StateManager


def fetch_issue_body(issue: Issue, cfg: RunConfig, state: StateManager) -> str:
    """Fetch an issue body from the remote platform, using cache if available.

    If the issue has a local description, use that instead of fetching.
    """
    # If issue has local description, use it directly
    if issue.description:
        return issue.description

    cached = state.get_cached_issue(issue.number)
    if cached is not None:
        return cached

    repo_cfg = cfg.repo_for_issue(issue)
    body = _fetch_from_platform(issue.number, repo_cfg.path, repo_cfg.platform)
    if body and not body.startswith("(Could not fetch"):
        state.cache_issue(issue.number, body)
    return body


def _fetch_from_platform(issue_number: int, repo_path: str, platform: str) -> str:
    """Fetch issue body from GitLab or GitHub."""
    try:
        if platform == "gitlab":
            result = subprocess.run(
                ["glab", "issue", "view", str(issue_number)],
                capture_output=True,
                text=True,
                cwd=repo_path,
                timeout=30,
                check=False,
            )
        elif platform == "github":
            result = subprocess.run(
                ["gh", "issue", "view", str(issue_number)],
                capture_output=True,
                text=True,
                cwd=repo_path,
                timeout=30,
                check=False,
            )
        else:
            return f"(Unknown platform '{platform}' for issue #{issue_number})"

        if result.returncode == 0:
            return result.stdout
        return (
            f"(Could not fetch issue #{issue_number} from {platform}. "
            f"Error: {result.stderr.strip()}. "
            f"Work from the title and context below.)"
        )
    except (subprocess.TimeoutExpired, FileNotFoundError) as e:
        return f"(Could not fetch issue #{issue_number}: {e}. Work from the title and context below.)"


def next_available_issue(cfg: RunConfig, completed: set[int],
                         in_progress: Optional[set[int]] = None) -> Optional[Issue]:
    """Find the next issue that can be assigned.

    Issues are sorted by wave then priority. An issue is available if:
    - Its status is "pending"
    - All its dependencies are in the completed set
    - It's not already in progress
    """
    if in_progress is None:
        in_progress = set()

    for issue in sorted(cfg.issues, key=lambda i: (i.wave, i.priority)):
        if issue.status == "pending" and issue.number not in in_progress:
            if all(dep in completed for dep in issue.depends_on):
                return issue
    return None


def get_in_progress_issues(cfg: RunConfig) -> set[int]:
    """Return set of issue numbers currently in progress."""
    return {i.number for i in cfg.issues if i.status == "in_progress"}


def get_pending_count(cfg: RunConfig) -> int:
    """Return count of pending + in_progress issues."""
    return sum(1 for i in cfg.issues if i.status in ("pending", "in_progress"))


def get_completed_count(cfg: RunConfig) -> int:
    """Return count of completed issues."""
    return sum(1 for i in cfg.issues if i.status == "completed")


def get_failed_count(cfg: RunConfig) -> int:
    """Return count of failed issues."""
    return sum(1 for i in cfg.issues if i.status == "failed")


def next_available_issue_global(
    configs: list[RunConfig],
    claimed_issues: Optional[set[tuple[str, int]]] = None,
) -> Optional[tuple[RunConfig, Issue]]:
    """Find the highest-priority available issue across ALL projects.

    Returns (cfg, issue) for the best available issue, or None.
    Priority: wave first, then issue priority, across all configs.

    claimed_issues: set of (config_path_str, issue_number) already claimed
    this cycle, to prevent double-assignment.
    """
    if claimed_issues is None:
        claimed_issues = set()

    best: Optional[tuple[RunConfig, Issue]] = None
    best_key = (999, 999)  # (wave, priority)

    for cfg in configs:
        completed = {i.number for i in cfg.issues if i.status == "completed"}
        in_progress = {i.number for i in cfg.issues if i.status == "in_progress"}
        cfg_path = str(cfg.config_path)

        # Exclude issues already claimed this cycle
        for claim_path, claim_num in claimed_issues:
            if claim_path == cfg_path:
                in_progress.add(claim_num)

        for issue in sorted(cfg.issues, key=lambda i: (i.wave, i.priority)):
            if issue.status != "pending" or issue.number in in_progress:
                continue
            if not all(dep in completed for dep in issue.depends_on):
                continue
            key = (issue.wave, issue.priority)
            if key < best_key:
                best = (cfg, issue)
                best_key = key
                break  # This config's best; compare with other configs

    return best


def next_retriable_issue_global(
    configs: list[RunConfig],
    claimed_issues: Optional[set[tuple[str, int]]] = None,
) -> Optional[tuple[RunConfig, Issue]]:
    """Find the best failed issue worth retrying across ALL projects.

    Prioritizes issues that block the most downstream work.
    Returns (cfg, issue) or None.
    """
    if claimed_issues is None:
        claimed_issues = set()

    best: Optional[tuple[RunConfig, Issue]] = None
    best_score = -1

    for cfg in configs:
        cfg_path = str(cfg.config_path)

        # Build map: issue -> how many issues depend on it (directly or transitively)
        dependents: dict[int, int] = {}
        for issue in cfg.issues:
            for dep in issue.depends_on:
                dependents[dep] = dependents.get(dep, 0) + 1

        for issue in cfg.issues:
            if issue.status != "failed":
                continue
            # Skip if already claimed
            if (cfg_path, issue.number) in claimed_issues:
                continue
            # Score: number of downstream dependents (higher = more critical)
            score = dependents.get(issue.number, 0)
            # Tiebreak: lower wave first, then lower priority number
            if score > best_score or (score == best_score and best is not None
                                       and (issue.wave, issue.priority)
                                       < (best[1].wave, best[1].priority)):
                best = (cfg, issue)
                best_score = score

    return best


def next_available_cross_project(
    cfg: RunConfig,
    exclude_config: Optional[str] = None,
    claimed_cross: Optional[set[tuple[str, int]]] = None,
) -> Optional[tuple["RunConfig", Issue]]:
    """Find the next available issue from any other project config.

    Returns (other_cfg, issue) or None. Skips the config at exclude_config path
    (the caller's own project). Respects claimed_cross to avoid duplicates.
    """
    from pathlib import Path
    from .config import load_config

    if claimed_cross is None:
        claimed_cross = set()

    config_dir = cfg.config_path.parent
    for config_path in sorted(config_dir.glob("*-issues.json")):
        resolved = config_path.resolve()
        if exclude_config and resolved == Path(exclude_config).resolve():
            continue
        try:
            other_cfg = load_config(config_path)
        except (SystemExit, Exception):
            continue

        completed = {i.number for i in other_cfg.issues if i.status == "completed"}
        in_progress = {i.number for i in other_cfg.issues if i.status == "in_progress"}
        # Also exclude issues already claimed this cycle
        for claim_path, claim_num in claimed_cross:
            if Path(claim_path).resolve() == resolved:
                in_progress.add(claim_num)
        issue = next_available_issue(other_cfg, completed, in_progress)
        if issue:
            return other_cfg, issue

    return None

File: internal/orchestrator/issues.go
package orchestrator

import (
	"context"
	"fmt"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// FetchIssueBody fetches an issue body from GitHub.
// If the issue has a local description, use that instead of fetching.
func FetchIssueBody(issue *Issue, cfg *RunConfig, state *StateManager) string {
	// If issue has local description, use it directly
	if issue.Description != "" {
		return issue.Description
	}

	repoCfg := cfg.RepoForIssue(issue)
	if repoCfg == nil {
		return fmt.Sprintf("(No repo configured for issue #%d)", issue.Number)
	}
	return fetchFromPlatform(issue.Number, repoCfg.Path, repoCfg.Platform)
}

func fetchFromPlatform(issueNumber int, repoPath, platform string) string {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	var cmd *exec.Cmd
	switch platform {
	case "gitlab":
		cmd = exec.CommandContext(ctx, "glab", "issue", "view", fmt.Sprintf("%d", issueNumber))
	case "github":
		cmd = exec.CommandContext(ctx, "gh", "issue", "view", fmt.Sprintf("%d", issueNumber))
	default:
		return fmt.Sprintf("(Unknown platform '%s' for issue #%d)", platform, issueNumber)
	}

	cmd.Dir = repoPath
	out, err := cmd.Output()
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			return fmt.Sprintf("(Could not fetch issue #%d from %s. Error: %s. Work from the title and context below.)",
				issueNumber, platform, strings.TrimSpace(string(exitErr.Stderr)))
		}
		return fmt.Sprintf("(Could not fetch issue #%d: %v. Work from the title and context below.)", issueNumber, err)
	}
	return string(out)
}

// NextAvailableIssue finds the next issue that can be assigned.
// Issues are sorted by wave then priority. An issue is available if:
// - Its status is "pending"
// - All its dependencies are in the completed set
// - It's not already in progress
func NextAvailableIssue(cfg *RunConfig, completed map[int]bool, inProgress map[int]bool) *Issue {
	if inProgress == nil {
		inProgress = make(map[int]bool)
	}

	// Sort issues by wave then priority
	issues := make([]*Issue, len(cfg.Issues))
	copy(issues, cfg.Issues)
	sort.Slice(issues, func(i, j int) bool {
		if issues[i].Wave != issues[j].Wave {
			return issues[i].Wave < issues[j].Wave
		}
		return issues[i].Priority < issues[j].Priority
	})

	for _, issue := range issues {
		if issue.Status != "pending" || inProgress[issue.Number] {
			continue
		}
		depsOK := true
		for _, dep := range issue.DependsOn {
			if !completed[dep] {
				depsOK = false
				break
			}
		}
		if depsOK {
			return issue
		}
	}
	return nil
}

// GetInProgressIssues returns set of issue numbers currently in progress.
func GetInProgressIssues(cfg *RunConfig) map[int]bool {
	result := make(map[int]bool)
	for _, i := range cfg.Issues {
		if i.Status == "in_progress" {
			result[i.Number] = true
		}
	}
	return result
}

// GetPendingCount returns count of pending issues only.
func GetPendingCount(cfg *RunConfig) int {
	count := 0
	for _, i := range cfg.Issues {
		if i.Status == "pending" {
			count++
		}
	}
	return count
}

// GetCompletedCount returns count of completed issues.
func GetCompletedCount(cfg *RunConfig) int {
	count := 0
	for _, i := range cfg.Issues {
		if i.Status == "completed" {
			count++
		}
	}
	return count
}

// GetFailedCount returns count of failed issues.
func GetFailedCount(cfg *RunConfig) int {
	count := 0
	for _, i := range cfg.Issues {
		if i.Status == "failed" {
			count++
		}
	}
	return count
}

// GetPRPendingCount returns count of issues with PRs waiting to be merged.
func GetPRPendingCount(cfg *RunConfig) int {
	count := 0
	for _, i := range cfg.Issues {
		if i.Status == "pr_pending" {
			count++
		}
	}
	return count
}

// GetPRPendingIssues returns issues that have open PRs waiting to be merged.
func GetPRPendingIssues(cfg *RunConfig) []*Issue {
	var issues []*Issue
	for _, i := range cfg.Issues {
		if i.Status == "pr_pending" {
			issues = append(issues, i)
		}
	}
	return issues
}

// NextRetriableIssue finds a failed issue worth retrying.
// Prioritizes issues that block the most downstream work.
func NextRetriableIssue(cfg *RunConfig, completed map[int]bool) *Issue {
	// Build map: issue -> how many pending issues depend on it
	dependents := make(map[int]int)
	for _, issue := range cfg.Issues {
		if issue.Status != "pending" {
			continue
		}
		for _, dep := range issue.DependsOn {
			dependents[dep]++
		}
	}

	var best *Issue
	bestScore := -1

	for _, issue := range cfg.Issues {
		if issue.Status != "failed" {
			continue
		}
		// Check if any pending issue depends on this one
		score := dependents[issue.Number]
		if score > bestScore || (score == bestScore && best != nil &&
			(issue.Wave < best.Wave || (issue.Wave == best.Wave && issue.Priority < best.Priority))) {
			best = issue
			bestScore = score
		}
	}

	// Only return if this issue actually blocks something
	if bestScore > 0 {
		return best
	}
	return nil
}

// ClaimedIssue represents a claimed issue by config path and issue number.
type ClaimedIssue struct {
	ConfigPath  string
	IssueNumber int
}

// NextAvailableIssueGlobal finds the highest-priority available issue across ALL projects.
// Returns (cfg, issue) for the best available issue, or nil.
func NextAvailableIssueGlobal(configs []*RunConfig, claimed []ClaimedIssue) (*RunConfig, *Issue) {
	claimedSet := make(map[ClaimedIssue]bool)
	for _, c := range claimed {
		claimedSet[c] = true
	}

	var best *Issue
	var bestCfg *RunConfig
	bestKey := [2]int{999, 999} // (wave, priority)

	for _, cfg := range configs {
		completed := make(map[int]bool)
		inProgress := make(map[int]bool)
		for _, i := range cfg.Issues {
			if i.Status == "completed" {
				completed[i.Number] = true
			} else if i.Status == "in_progress" {
				inProgress[i.Number] = true
			}
		}

		// Exclude issues already claimed this cycle
		for claim := range claimedSet {
			if claim.ConfigPath == cfg.ConfigPath {
				inProgress[claim.IssueNumber] = true
			}
		}

		// Sort issues
		issues := make([]*Issue, len(cfg.Issues))
		copy(issues, cfg.Issues)
		sort.Slice(issues, func(i, j int) bool {
			if issues[i].Wave != issues[j].Wave {
				return issues[i].Wave < issues[j].Wave
			}
			return issues[i].Priority < issues[j].Priority
		})

		for _, issue := range issues {
			if issue.Status != "pending" || inProgress[issue.Number] {
				continue
			}
			depsOK := true
			for _, dep := range issue.DependsOn {
				if !completed[dep] {
					depsOK = false
					break
				}
			}
			if !depsOK {
				continue
			}

			key := [2]int{issue.Wave, issue.Priority}
			if key[0] < bestKey[0] || (key[0] == bestKey[0] && key[1] < bestKey[1]) {
				best = issue
				bestCfg = cfg
				bestKey = key
				break // This config's best; compare with other configs
			}
		}
	}

	if best == nil {
		return nil, nil
	}
	return bestCfg, best
}

// NextRetriableIssueGlobal finds the best failed issue worth retrying across ALL projects.
// Prioritizes issues that block the most downstream work.
func NextRetriableIssueGlobal(configs []*RunConfig, claimed []ClaimedIssue) (*RunConfig, *Issue) {
	claimedSet := make(map[ClaimedIssue]bool)
	for _, c := range claimed {
		claimedSet[c] = true
	}

	var best *Issue
	var bestCfg *RunConfig
	bestScore := -1

	for _, cfg := range configs {
		// Build map: issue -> how many issues depend on it
		dependents := make(map[int]int)
		for _, issue := range cfg.Issues {
			for _, dep := range issue.DependsOn {
				dependents[dep]++
			}
		}

		for _, issue := range cfg.Issues {
			if issue.Status != "failed" {
				continue
			}
			// Skip if already claimed
			claim := ClaimedIssue{ConfigPath: cfg.ConfigPath, IssueNumber: issue.Number}
			if claimedSet[claim] {
				continue
			}

			score := dependents[issue.Number]
			// Tiebreak: lower wave first, then lower priority number
			if score > bestScore || (score == bestScore && best != nil &&
				(issue.Wave < best.Wave || (issue.Wave == best.Wave && issue.Priority < best.Priority))) {
				best = issue
				bestCfg = cfg
				bestScore = score
			}
		}
	}

	if best == nil {
		return nil, nil
	}
	return bestCfg, best
}

// NextAvailableCrossProject finds the next available issue from any other project config.
// Returns (other_cfg, issue) or nil. Skips the config at excludeConfig path.
func NextAvailableCrossProject(cfg *RunConfig, excludeConfig string, claimed []ClaimedIssue) (*RunConfig, *Issue) {
	configDir := filepath.Dir(cfg.ConfigPath)
	matches, err := filepath.Glob(filepath.Join(configDir, "*-issues.json"))
	if err != nil {
		return nil, nil
	}
	sort.Strings(matches)

	claimedSet := make(map[ClaimedIssue]bool)
	for _, c := range claimed {
		claimedSet[c] = true
	}

	for _, configPath := range matches {
		resolved, _ := filepath.Abs(configPath)
		if excludeConfig != "" {
			excludeResolved, _ := filepath.Abs(excludeConfig)
			if resolved == excludeResolved {
				continue
			}
		}

		otherCfg, err := LoadConfig(configPath)
		if err != nil {
			continue
		}

		completed := make(map[int]bool)
		inProgress := make(map[int]bool)
		for _, i := range otherCfg.Issues {
			if i.Status == "completed" {
				completed[i.Number] = true
			} else if i.Status == "in_progress" {
				inProgress[i.Number] = true
			}
		}

		// Also exclude issues already claimed this cycle
		for claim := range claimedSet {
			claimResolved, _ := filepath.Abs(claim.ConfigPath)
			if claimResolved == resolved {
				inProgress[claim.IssueNumber] = true
			}
		}

		issue := NextAvailableIssue(otherCfg, completed, inProgress)
		if issue != nil {
			return otherCfg, issue
		}
	}

	return nil, nil
}
